sight_service/message_queue.py

```diff
@@ -0,0 +1,255 @@
+"""A message queue implementation using reader-writer locks."""
+
+import abc
+import copy
+from typing import Callable, Dict, Generic, Optional, Protocol, TypeVar
+import uuid
+
+from overrides import overrides
+from readerwriterlock import rwlock
+
+
+class UUIDStrategy(abc.ABC):
+
+  @abc.abstractmethod
+  def generate_id(self) -> int:
+    pass
+
+
+class IncrementalUUID(UUIDStrategy):
+
+  def __init__(self):
+    self.current_id = 1
+
+  def generate_id(self) -> int:
+    unique_id = self.current_id
+    self.current_id += 1
+    return unique_id
+
+
+class RandomUUID(UUIDStrategy):
+
+  def generate_id(self) -> int:
+    return uuid.uuid4().int  # Using the integer representation of UUID
+
+
+# Define a generic type variable for messages
+T = TypeVar('T')
+
+
+class IMessageQueue(Protocol, Generic[T]):
+  """A message queue is a data structure that stores messages."""
+
+  def push_message(self, message: T) -> int:
+    """Pushes a message to the queue.
+
+    Args:
+      message: The message to push.
+    """
+    ...
+
+  def process_messages(
+      self, worker_id: str, new_batch_size: Optional[int] = None
+      ) -> Dict[int, T]:
+    """Processes a batch of messages for a given worker.
+
+    Args:
+      worker_id: The ID of the worker that will process the messages.
+      new_batch_size: The size of the batch to process. If not provided, the
+        default batch size will be used.
+    """
+    ...
+
+  def complete_message(
+      self, message_id: str, worker_id: str, extra_details: Optional[any] = None
+    ) -> None:
+    """Completes a message of the given message ID of the given worker it moves it to the completed queue.
+
+    Args:
+      message_id: The ID of the message to complete.
+      worker_id: The ID of the worker that completed the message.
+      extra_details: Any extra details to add to the completed message. (only
+        for dict types messages)
+    """
+    ...
+
+  def get_status(self) -> Dict[str, int]:
+    """Returns the status of the message queue."""
+    ...
+
+  def get_all_messages(self) -> Dict[str, Dict[int, T]]:
```

Comment:
The return type of this method is not right since the three queues have different returns types:
    self.pending: Dict[int, T] = {}
    self.active: Dict[str, Dict[int, T]] = {}
    self.completed: Dict[int, T] = {}

Instead, I'm noticing that you normally use this method to get just one sub-class of messages, not actually all of them. Can this be 3 separate functions for the get_pending(), get_active() and get_completed()?

Reply:
yes , I was thinking the same as for the need of only one type of list , I was returning all three copied lists

Reply:
will update

Reply:
updated